File: Baekjoon/Class4/18.13549 숨바꼭질 3.py
from collections import deque
import sys
input = sys.stdin.readline

n, k = map(int, input().split())
depth_limit = max(n, k) * 2

dq = deque() #0-1 너비 우선 탐색
depths = [-1] * (depth_limit + 1)
dq.append(n) #시작
depths[n] = 0

while dq:
    start = dq.popleft()
    if start == k:
        break

    end = start * 2
    if end <= depth_limit and depths[end] == -1: #순간 이동
        dq.appendleft(end)
        depths[end] = depths[start]

    for end in (start - 1, start + 1): #걷기
        if end >= 0 and end <= depth_limit and depths[end] == -1:
            dq.append(end)
            depths[end] = depths[start] + 1

print(depths[k])

File: Baekjoon/Class4/18.13549 숨바꼭질 3.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim();

const [n, k] = input.split(" ").map(Number);
const depthLimit = Math.max(n, k) * 2;

const bfs = [n]; // 0-1 너비 우선 탐색
const depths = new Array(depthLimit + 1).fill(-1);
depths[n] = 0; // 시작

while(bfs.length !== 0) {
    const start = bfs.pop();
    if(start === k) break;

    const teleport = start * 2;
    if(teleport <= depthLimit && depths[teleport] === -1) { // 순간 이동
        bfs.push(teleport);
        depths[teleport] = depths[start];
    }

    for(const walk of [start - 1, start + 1]) // 걷기
        if(walk >= 0 && walk <= depthLimit && depths[walk] === -1) {
            bfs.unshift(walk);
            depths[walk] = depths[start] + 1;
        }
}

console.log(depths[k]);